File: September 2025 GFG SOLUTION/September-01.py
class Solution:
    def sumOfModes(self, arr, k):
        n, sum_val = len(arr), 0
        freq = defaultdict(int)
        buckets = defaultdict(set)
        
        for i in range(k):
            freq[arr[i]] += 1
        
        for val, f in freq.items():
            buckets[f].add(val)
        
        max_freq = max(buckets.keys())
        sum_val += min(buckets[max_freq])
        
        for i in range(k, n):
            out, in_val = arr[i - k], arr[i]
            
            buckets[freq[out]].remove(out)
            if not buckets[freq[out]]:
                del buckets[freq[out]]
            
            freq[out] -= 1
            if freq[out] > 0:
                buckets[freq[out]].add(out)
            else:
                del freq[out]
            
            if freq[in_val] > 0:
                buckets[freq[in_val]].remove(in_val)
                if not buckets[freq[in_val]]:
                    del buckets[freq[in_val]]
            
            freq[in_val] += 1
            buckets[freq[in_val]].add(in_val)
            
            max_freq = max(buckets.keys())
            sum_val += min(buckets[max_freq])
        
        return sum_val

File: September 2025 GFG SOLUTION/September-01.java
class Solution {
    public int sumOfModes(int[] arr, int k) {
        int n = arr.length, sum = 0;
        Map<Integer, Integer> freq = new HashMap<>();
        TreeMap<Integer, TreeSet<Integer>> buckets = new TreeMap<>();
        
        for (int i = 0; i < k; i++) freq.merge(arr[i], 1, Integer::sum);
        freq.forEach((val, f) -> buckets.computeIfAbsent(f, x -> new TreeSet<>()).add(val));
        sum += buckets.lastEntry().getValue().first();
        
        for (int i = k; i < n; i++) {
            int out = arr[i - k], in = arr[i];
            
            TreeSet<Integer> outSet = buckets.get(freq.get(out));
            outSet.remove(out);
            if (outSet.isEmpty()) buckets.remove(freq.get(out));
            
            freq.merge(out, -1, Integer::sum);
            if (freq.get(out) > 0) 
                buckets.computeIfAbsent(freq.get(out), x -> new TreeSet<>()).add(out);
            else freq.remove(out);
            
            if (freq.containsKey(in)) {
                TreeSet<Integer> inSet = buckets.get(freq.get(in));
                inSet.remove(in);
                if (inSet.isEmpty()) buckets.remove(freq.get(in));
            }
            
            freq.merge(in, 1, Integer::sum);
            buckets.computeIfAbsent(freq.get(in), x -> new TreeSet<>()).add(in);
            sum += buckets.lastEntry().getValue().first();
        }
        return sum;
    }
}

